go/libraries/doltcore/doltdb/local_table_editor.go

```diff
@@ -0,0 +1,309 @@
+			}
+			if !ok {
+				return nil
+			}
+			err = lte.deleteReferencedRows(ctx, dRow)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return lte.tableEditor.DeleteKey(ctx, key)
+}
+
+// DeleteRow removes the given row from the table, along with any applicable rows from tables that have a foreign key
+// referencing this table.
+func (lte *LocalTableEditor) DeleteRow(ctx context.Context, dRow row.Row) error {
+	globalTableEditor.writeMutex.RLock()
+	defer globalTableEditor.writeMutex.RUnlock()
+
+	return lte.deleteRow(ctx, dRow)
+}
+
+// UpdateRow takes the current row and new row, and updates it accordingly. Any applicable rows from tables that have a
+// foreign key referencing this table will also be updated.
+func (lte *LocalTableEditor) UpdateRow(ctx context.Context, dOldRow row.Row, dNewRow row.Row) error {
+	globalTableEditor.writeMutex.RLock()
+	defer globalTableEditor.writeMutex.RUnlock()
+
+	return lte.updateRow(ctx, dOldRow, dNewRow, true)
+}
+
+func (lte *LocalTableEditor) deleteReferencedRows(ctx context.Context, dRow row.Row) error {
+	if len(lte.tablesToUpdate) > 0 && lte.restrictedDelete {
+		return fmt.Errorf("delete operations are restricted on this table due to a foreign key reference option")
+	}
+
+	for _, foreignKey := range lte.tablesToUpdate {
+		referencedLte, ok := globalTableEditor.tables[foreignKey.TableName]
+		if !ok {
+			var err error
+			referencedLte, err = globalTableEditor.getTableEditorOverride(ctx, foreignKey.TableName, nil, nil)
+			if err != nil {
+				return err
+			}
+		}
+		indexKey, err := lte.reduceRowAndConvert(lte.tableEditor.nbf, foreignKey.ReferencedTableColumns, foreignKey.TableColumns, dRow)
+		if err != nil {
+			return err
+		}
+		relevantRows, err := referencedLte.tableEditor.GetIndexedRows(ctx, indexKey, foreignKey.TableIndex)
+		if err != nil {
+			return err
+		}
+
+		switch foreignKey.OnDelete {
+		case ForeignKeyReferenceOption_Cascade:
+			for _, rowToDelete := range relevantRows {
+				err = referencedLte.DeleteRow(ctx, rowToDelete)
+				if err != nil {
+					return err
+				}
+			}
+		case ForeignKeyReferenceOption_SetNull:
+			for _, oldRow := range relevantRows {
+				newRow := oldRow
+				for _, colTag := range foreignKey.TableColumns {
+					newRow, err = newRow.SetColVal(colTag, types.NullValue, referencedLte.tableEditor.tSch)
+					if err != nil {
+						return err
+					}
+				}
+				err = referencedLte.updateRow(ctx, oldRow, newRow, false)
+				if err != nil {
+					return err
+				}
+			}
+		default:
+			//DefaultAction, NoAction, and Restrict will cause restrictedDelete to be set, so it's not one of those
+			return fmt.Errorf("unknown ON DELETE reference option on `%s`: `%s`", foreignKey.Name, foreignKey.OnDelete.String())
+		}
+	}
+
+	return nil
+}
+
+// deleteRow is the same as DeleteRow, except that it does not acquire any locks.
+func (lte *LocalTableEditor) deleteRow(ctx context.Context, dRow row.Row) error {
+	if !globalTableEditor.DisableForeignKeyChecks {
```

Comment:
This check belongs in the method being called, not here. Always call the method, which sometimes does nothing and returns nil.